File: Iteration_2/src/main/java/CourseSection.java
import java.sql.Time;
import java.util.ArrayList;

public class CourseSection {
   private String sectionID;
   private ArrayList<TimeSlot> timeSlots;
   private ArrayList<Student> waitList;
   private int capacity;
   private Lecturer lecturer;
   private Course parentCourse;
   private ArrayList<Student> currentStudents;

   // constructors
   public CourseSection() {
      this.timeSlots = new ArrayList<TimeSlot>();
      this.currentStudents = new ArrayList<Student>();
      this.waitList = new ArrayList<Student>();
   }

   public CourseSection(String sectionID, int capacity, Course parentCourse) {
      this();
      this.capacity = capacity;
      this.sectionID = sectionID;
      this.parentCourse = parentCourse;
   }

   public CourseSection(String sectionId, int capacity, Lecturer lecturer) {
      this();
      this.sectionID = sectionId;
      this.capacity = capacity;
      this.lecturer = lecturer;
   }

   public CourseSection(String sectionId, int capacity) {
      this();
      this.sectionID = sectionId;
      this.capacity = capacity;
   }

   public String getSectionID() {
      return this.sectionID;
   }

   public void setSectionID(String sectionID) {
      this.sectionID = sectionID;
   }

   public ArrayList<TimeSlot> getTimeSlots() {
      return timeSlots;
   }

   public void setTimeSlot(ArrayList<TimeSlot> timeSlots) {
      this.timeSlots = timeSlots;
   }

   public ArrayList<Student> getWaitList() {
      return waitList;
   }

   public void setWaitList(ArrayList<Student> waitList) {
      this.waitList = waitList;
   }

   public Lecturer getLecturer() {
      return lecturer;
   }

   public void setLecturer(Lecturer lecturer) {
      this.lecturer = lecturer;
   }

   public int getCapacity() {
      return this.capacity;
   }

   public void setCapacity(int capacity) {
      this.capacity = capacity;
   }

   public Course getParentCourse() {
      return parentCourse;
   }

   public void setParentCourse(Course parentCourse) {
      this.parentCourse = parentCourse;
   }

   public ArrayList<Student> getCurrentStudents() {
      return currentStudents;
   }

   public void addStudentToSection(Student currentStudent) {
      this.currentStudents.add(currentStudent);
   }

}

File: Iteration_3/CourseSection.py
from typing import List, Optional
from Lecturer import Lecturer
from Course import Course
from TimeSlot import TimeSlot

class CourseSection:
    def __init__(self, capacity: int = None, parent_course: 'Course' = None, lecturer: 'Lecturer' = None, section_id: str = None):
        self.__capacity = capacity
        self.__parent_course = parent_course
        self.__lecturer = lecturer
        self.__current_students: List['Student'] = []
        self.section_id = section_id
        self.time_slots: List['TimeSlot'] = []  # Initialize time_slots
        self.wait_list: List['Student'] = []    # Initialize wait_list

    def get_capacity(self) -> int:
        return self.__capacity

    def set_capacity(self, capacity: int) -> None:
        self.__capacity = capacity

    def get_parent_course(self) -> Optional['Course']:
        return self.__parent_course

    def set_parent_course(self, parent_course: 'Course') -> None:
        self.__parent_course = parent_course

    def get_current_students(self) -> List['Student']:
        return self.__current_students

    def add_student_to_section(self, student: 'Student') -> None:
        self.__current_students.append(student)

    def get_lecturer(self) -> Optional['Lecturer']:
        return self.__lecturer
    
    def set_lecturer(self, lecturer: 'Lecturer') -> None:
        self.__lecturer = lecturer

    def get_section_id(self) -> str:
        return self.section_id

    def set_section_id(self, section_id: str) -> None:
        self.section_id = section_id

    def get_time_slots(self) -> List['TimeSlot']:
        return self.time_slots

    def set_time_slots(self, time_slots: List['TimeSlot']) -> None:
        self.time_slots = time_slots

    def get_wait_list(self) -> List['Student']:
        return self.wait_list

    def set_wait_list(self, wait_list: List['Student']) -> None:
        self.wait_list = wait_list

    def has_capacity(self) -> bool:
        return len(self.__current_students) < self.__capacity